Swift AccessUtils: improvements and bug fixes

eeckstein opened the pull request:
While I was using the new AccessUtils for a new optimization pass I discovered some areas for improvements.
Changes:

* AccessBase: remove the unhealthy redundancy between `kind` and `baseAddress` types. Now AccessBase is single enum with the relevant base objects/addresses as payloads.

* AccessBase: for `global`, store the `GlobalValue` and not a `global_address` instruction, which is more accurate (because there can be multiple `global_addr`s for a single global variable)

* AccessBase: drop the support for function argument "pointers". The `pointer` is now always a `pointer_to_address` instruction. This also simplifies `PointerIdentification`: either it finds a matching `address_to_pointer` or it bails.

* AccessBase: improve `func isDistinct(from:)`. There are more possibilities to prove that two access bases do not alias.

* AccessBase: replace `var isUniquelyIdentified` with `var hasKnownStorageKind` which is more useful for aliasing checking.

* AccessPath: fix `func isDistinct(from:)`. `SmallProjectionPath.matches` is the wrong way to check if two expression paths may overlap. Instead use the new `SmallProjectionPath.mayOverlap`.

* AccessStoragePathWalker: rename `getAccessStorage` -> `visitAccessStorageRoots` and let it return false if it's not a class/reference AccessBase.

* add tests for `AccessPath.isDistinct(from:)`

eeckstein committed:
SmallProjectionPath: add `mayOverlap(with:)`

"Overlapping" means that both paths may project the same field.

eeckstein committed:
Swift SIL: add `NominalFieldsArray.getNameOfField(withIndex:)`

eeckstein commented:
@swift-ci smoke test

atrick commented:
> While I was using the new AccessUtils for a new optimization pass I discovered some areas for improvements.
Changes:

Thank you for clearly summarizing the changes!

> AccessBase: remove the unhealthy redundancy between kind and baseAddress types. Now AccessBase is single enum with the relevant base objects/addresses as payloads.

Looks good to me at first glance.

> AccessBase: for global, store the GlobalValue and not a global_address instruction, which is more accurate (because there can be multiple global_addrs for a single global variable)

That's right. But more importantly the base of global variable is not necessarilly a global_addr. We need to handle addressors!

> AccessBase: Similarly, for class, tail and box use the underlying class/box reference value + the field index as source of truth. Still, store a "representative" ref_element_addr/ref_tail_addr/project_box as payload (because it's easier), but use the underlying reference value + field indices to define equality.

AccessBase should not have a custom equality operator. The definition of equality on values is that they are substitutable. Two AccessBase values with different ref_element_addr are *not* substitutable. The first implementation of MemAccessUtils did that and was broken.

Given an AccessBase, it must be possible to rewrite all those accesses just from the base and projection path. If we substitute values with different base addresses, then that breaks.

ref_element_addr *is* the base address. And unlike address projections, reference-to-address conversions are non-fungible. You cannot substitute one for the other. They are not rematerializable from their object and field index.

Note that we could try to make ref_element_addr rematerializable by eliminating the isImmutable flag. But that would not be useful because we still wouldn't be able to "see-through" borrow scopes. We could also eliminate those spurious borrow scopes. But there are still copies and all sorts of arbitrary ownership casts. The more important point is that alias analysis should always find the storage root before doing a comparison.

For points-to alias analysis, you should be comparing the "root" of the access, not the base. The root identifies the object being accessed. A different API should be used that explicitly says what it does instead of magically hiding the base address and pretending the base is a root.

Some of our earlier designs had a separate AccessRoot abstraction. It should be possible to query the root of an AccessBase on-the-fly. But it's also be useful to encapsulate AccessRoot in a type alongside the base. You don't want to redo a traversal across phis on every comparison. AccessRoot is also needed to use the identity of the storage as a hash key instead of the base address. In fact, we need that for exclusivity optimization!

To recap: If you want to compare the storage of an access, then you should always compute the storage root first. It doesn't make sense to treat the ref_element_addr operand as a "fake root".

> AccessBase: for pointer always store a (non-address) value of type Builtin.RawPointer - whatever that might be: either the operand of an address_to_pointer, a function argument or any other root value of the RawPointer.

The pointer_to_address *is* the base address, and it is also not substitutable. It carries the information about the pointer itself. Includng the information about aliasing. In the future, it will carry more information.

The *only* time AccessBase can "see through" a pointer_to_address is when all pointer values originate from a single common address *which has the same type as the access*.

For type based alias analysis, it's critical to retain the pointer_to_address because it carries the type information for the access.

Of course, points-to alias analysis needs to find the pointer's provenance. That is why we need APIs that give you the storage root, not the base address!

> AccessBase: improve func isDistinct(from:). There are more possibilities to prove that two access bases do not alias.

Yeah. This is great, but for alias analysis we should be using the storage root for box, class, tail, and pointer. Maybe this API should be moved to an AccessRoot type instead.

> AccessBase: replace var isUniquelyIdentified with var hasKnownStorageKind which is more useful for aliasing checking.

isUniquelyIdentified is a fundamental property of a storage root. All of our known-use analyses are based on it. It's the only way to prove no aliasing without information about the other accesses. It can probably be implemented based on isDifferentAllocation and hasKnownStorageKind, but the exclusivity information is missing.

hasKnownStorageKind seems fine but drops information about argument exclusivity.

> AccessPath: fix func isDistinct(from:). SmallProjectionPath.matches is the wrong way to check if two expression paths may overlap. Instead use the new SmallProjectionPath.mayOverlap.

Yeah, this is tricky. Sometimes we need to build or compare the projection paths in the opposite direction. AccessSummaryAnalysis needs to push a component for each address projection instead of popping. That was one of the first uses of the AccessPath trie.

> simplify PointerIdentification: always return a value of type Builtin.RawPointer. This makes the enum AddressOrPointerArgument obsolete.

To compute AccessBase, PointerIdentification needs to find a single common address or just abort. You can't remove the type check, it is needed for correctness.  I don't remember why this utility needed to return a FunctionArgument. That doesn't make sense to me now.

What you implemented was a walk to find the root of a pointer. That's useful for points-to alias analysis where you don't care about types.

You could possibly combine these implementations but you can't change the way we search for a base address.

eeckstein committed:
Swift AccessUtils: improvements and bug fixes

While I was using the new AccessUtils for a new optimization pass I discovered some areas for improvements. Also I found some bugs.

Changes:

* AccessBase: remove the unhealthy redundancy between `kind` and `baseAddress` types. Now AccessBase is single enum with the relevant base objects/addresses as payloads.

* AccessBase: for `global`, store the `GlobalValue` and not a `global_address` instruction, which is more accurate (because there can be multiple `global_addr`s for a single global variable)

* AccessBase: drop the support for function argument "pointers". The `pointer` is now always a `pointer_to_address` instruction. This also simplifies `PointerIdentification`: either it finds a matching `address_to_pointer` or it bails.

* AccessBase: improve `func isDistinct(from:)`. There are more possibilities to prove that two access bases do not alias.

* AccessBase: replace `var isUniquelyIdentified` with `var hasKnownStorageKind` which is more useful for aliasing checking.

* AccessPath: fix `func isDistinct(from:)`. `SmallProjectionPath.matches` is the wrong way to check if two expression paths may overlap. Instead use the new `SmallProjectionPath.mayOverlap`.

* AccessStoragePathWalker: rename `getAccessStorage` -> `visitAccessStorageRoots` and let it return false if it's not a class/reference AccessBase.

* add tests for `AccessPath.isDistinct(from:)`

eeckstein commented:
@atrick 
> AccessBase should not have a custom equality operator. 

okay, got it. I removed it.

> To compute AccessBase, PointerIdentification needs to find a single common address or just abort. You can't remove the type check, it is needed for correctness. I don't remember why this utility needed to return a FunctionArgument. That doesn't make sense to me now.

So either we look through an `address_to_pointer` - `pointer_to_address` pair or we stop at the `pointer_to_address`. That totally makes sense. I was confused by this weird handling of function arguments.
I changed it.

I'm wondering if we even need this in access utils or can just rely on SILCombine to simplify such pairs.

> isUniquelyIdentified is a fundamental property of a storage root.

The problem with `isUniquelyIdentified` is that it's not well defined without a corresponding equality operation. Its comment
```
  /// Returns `true` if we can reliably compare this `AccessBase`
  /// with another `AccessBase` for equality.
```
is not correct, assuming the standard equality operator. Simplest example:
```
  %1 = alloc_ref $X
  %access1 = ref_element_addr %1 : $X, #X.f
  %access2 = ref_element_addr %1 : $X, #X.f
```
`%access` and `%access2` are "uniquely identified". They are not equal but may alias.

So, every client of `isUniquelyIdentified` needs to understand its internal implementation to be able to use it correctly. This is not a good API.

Instead, the `isDistinct(from:)` combines both, the uniqueness check and the comparison. And this solves the problem.

eeckstein commented:
@swift-ci smoke test

Angelogeb commented:
> I was confused by this weird handling of function arguments.

I think sometimes it's valuable to know that a pointer originates from a function argument, for example a caller might have been the one producing that pointer.

> AccessBase: remove the unhealthy redundancy between kind and baseAddress types. Now AccessBase is single enum with the relevant base objects/addresses as payloads.

This is indeed nicer than before and allows to more easily unpack a AddressBase without kind/class checks. The previous version was more "open".  But there is still no good support for addressors (for global_init the base currently is the pointer_to_address instruction).

eeckstein commented:
> I think sometimes it's valuable to know that a pointer originates from a function argument, for example a caller might have been the one producing that pointer.

Sure, but I don't think this belongs to the access utils. If a client needs this information it can use AccessBase to get the `pointer_to_address` and do its own use-def walk to find out if it comes from a function argument.
